### DISCUSSION TITLE
What is the practical limitation to PDF size?

### DISCUSSION THREAD

**Discussion by tpluscode** [inception-project/inception#4588] [Q&A]
I have some large PDFs I wish to upload and annotate. They are mostly imagery which causes the large size. In the issue #4587 I used a 25 MB document which is already fairly slow to render the pages initially and when the page layout shifts. Thinking about this, is the rendering does client-side?

**Comment by reckart** [comment]
Documents are rendered client-side using [PDF.js](https://mozilla.github.io/pdf.js/) (a version from the 2.x series).

**Comment by reckart** [comment]
So I tried the PDF you attached to the issue - it takes a moment to render, but at least from my perspective, it seems ok - and faster on my machine than it seems to have been in your video (Chrome 122.0.6261.94, macOS Ventura 13.6.3, Apple M2).

So far, I did not observe issues with PDF rendering speed (although your PDF is indeed pushing it). Large PDFs (we talk e.g. books with a hundreds of pages) can slow things down because INCEpTION then has to deal with a very large annotation file and saving that can take a while (INCEpITON uses a write-through cache).

**Comment by tpluscode** [comment]
I do not mind the speed. Especially that the video was captured of a fairly old machine. But I do have several much larger PDFs and was wondering how far I can push it.

**Comment by reckart** [comment]
If your PDF is large mostly because of images, then we are looking at the rendering speed in the client as a bottleneck.

If your PDF is large mostly because of text, then the annotation data file managed by INCEpTION is likely the bottleneck. In this case, updating annotations would be slow.
